### PR TITLE
Error descriptions

### PR DESCRIPTION
**Pull request by AndreiEvstratenko**
# Pull Request

## Description

- Add description for every error. A bit refactoring

### Motivation and Context

- closes #1839

### Types of changes

- [ ] Bug fix
- [ ] New feature
- [x] Maintenance

### CODE REVIEW

**Comment by ox** [comment]
At the very least the failure callbacks for these methods should get an error. That way the developer can figure out how to handle those errors in their application (showing an alert, sending a metric, retrying, etc...).

**Comment by brentchow** [comment]
Wouldn't a developer expect a standard Error type like [`NSError`](https://developer.apple.com/documentation/foundation/nserror) instead of a custom `SvrfError`

**Comment by brentchow** [comment]
based on our convo in slack, @AndreiEvstratenko is going to make success/error closures optional.

File: SvrfSDK/Source/SvrfError.swift [hunk 1]
```diff
@@ -3,24 +3,51 @@
...
+    
+    enum Media: String {
+        case Response = "Svrf: Media. Server response error."
+        case ResponseNoMedia = "Svrf: Media. There is no media in the server response."
+    }
     
     enum GetNode: String {
-        case IncorrectMediaType = "Svrf: Incorrect media type. Media type should be equal _3d"
-        case GetScene = "Svrf: Can't get scene from media. Scene is equal nil"
+        case IncorrectMediaType = "Svrf: GetNode. Incorrect media type. Media type should be equal _3d."
```
Comment (brentchow) [inline SvrfSDK/Source/SvrfError.swift]:
```suggestion
        case IncorrectMediaType = "Svrf: GetNode. Incorrect media type. Media type should equal _3d."
```

File: SvrfSDK/Source/SvrfSDK.swift [hunk 2]
```diff
@@ -142,38 +150,37 @@ public class SvrfSDK: NSObject {
...
-       - error: Error message.
+     - type: The type(s) of *Media* to be returned (comma separated).
+     - stereoscopicType: Search only for *Media* with a particular stereoscopic type.
+     - category: Search only for *Media* with a particular category.
+     - size: The number of results to return per-page, from 1 to 100.
+     - nextPageCursor: Pass this cursor ID to get the next page of results.
+     - onSuccess: Success closure.
+     - mediaArray: An array of *Media* from the Svrf API.
+     - onFailure: Error closure.
+     - error: Error message.
```
Comment (brentchow) [inline SvrfSDK/Source/SvrfSDK.swift]:
```suggestion
     - error: A SvrfError.
```

File: SvrfSDK/Source/SvrfSDK.swift [hunk 3]
```diff
@@ -184,28 +191,27 @@ public class SvrfSDK: NSObject {
...
-       - id: ID of *Media* to fetch.
-       - onSuccess: Success closure.
-       - media: *Media* from the Svrf API.
-       - onFailure: Error closure.
-       - error: Error message.
+     - id: ID of *Media* to fetch.
+     - onSuccess: Success closure.
+     - media: *Media* from the Svrf API.
+     - onFailure: Error closure.
+     - error: Error message.
```
Comment (brentchow) [inline SvrfSDK/Source/SvrfSDK.swift]:
```suggestion
     - error: A SvrfError.
```

File: SvrfSDK/Source/SvrfError.swift [hunk 4]
```diff
@@ -3,24 +3,51 @@
...
-    case Plist = "SVRF: Problems with .plist file"
-    case ApiKey = "SVRF: Problems with API key. Please, set your API key into .plist file for SVRF_API_KEY field"
-    case AuthResponse = "SVRF: Authentication response error"
-    case SearchResponse = "SVRF: Search response error"
-    case TrendingResponse = "SVRF: Trending response error"
-    case MediaResponse = "SVRF: Media response error"
-    case GetScene = "SVRF: Get scene error"
-    case CreateScene = "SVRF: Create scene error"
+    enum Search: String {
+        case Response = "Svrf: Search. Server response error."
```
Comment (brentchow) [inline SvrfSDK/Source/SvrfError.swift]: I think error messages Don't need the name Svrf or the method name. So they could all just say `Server response error`. The developer can add any extra data/description.

File: SvrfSDK/Source/SvrfSDK.swift [hunk 5]
```diff
@@ -217,22 +223,20 @@ public class SvrfSDK: NSObject {
      
      - attention: Face filters should be retrieved using the `getFaceFilter` method.
      - parameters:
-       - media: The *Media* to generate the *SCNNode* from. The *type* must be `_3d`.
+     - media: The *Media* to generate the *SCNNode* from. The *type* must be `_3d`.
```
Comment (brentchow) [inline SvrfSDK/Source/SvrfSDK.swift]: Update documentation to include: 

```
- onSuccess: Success closure.
- media:  The *Media* to generate the *SCNNode* from. The *type* must be `_3d`.
- onFailure: Failure closure.
- error: A SvrfError.
```

File: SvrfSDK/Source/SvrfSDK.swift [hunk 6]
```diff
@@ -217,22 +223,20 @@ public class SvrfSDK: NSObject {
...
      - returns: SCNNode?
      */
-    public static func getNodeFromMedia(media: Media) -> SCNNode? {
+    public static func getNodeFromMedia(media: Media, onSuccess success: @escaping (_ node: SCNNode) -> Void, onFailure failure: ((_ error: SvrfError) -> Void)? = nil) {
```
Comment (brentchow) [inline SvrfSDK/Source/SvrfSDK.swift]: Update README.md example with these changes

File: SvrfSDK/Source/SvrfSDK.swift [hunk 7]
```diff
@@ -254,141 +258,141 @@ public class SvrfSDK: NSObject {
...
         }
     }
-    
+
     /**
      The SVRF API allows you to access all of SVRF's ARKit compatible face filters and stream them directly to your app. Use the `getFaceFilter` method to stream a face filter to your app and convert it into a *SCNNode* in runtime.
      
      - parameters:
-       - media: The *Media* to generate the face filter from. The *type* must be `_3d`.
+     - media: The *Media* to generate the face filter from. The *type* must be `_3d`.
```
Comment (brentchow) [inline SvrfSDK/Source/SvrfSDK.swift]: Update documentation to include:

```
- onSuccess: Success closure.
- media:  The *Media* to generate the face filter from. The *type* must be `_3d`.
- onFailure: Failure closure.
- error: A SvrfError.
```

File: SvrfSDK/Source/SvrfSDK.swift [hunk 8]
```diff
@@ -254,141 +258,141 @@ public class SvrfSDK: NSObject {
...
      - returns: SCNNode
      */
-    public static func getFaceFilter(media: Media) -> SCNNode {
-        let faceFilter = SCNNode()
-        
-        if media.type == ._3d, let glbUrlString = media.files?.glb, let glbUrl = URL(string: glbUrlString) {
-            let modelSource = GLTFSceneSource(url: glbUrl)
+    public static func getFaceFilter(with device: MTLDevice, media: Media, onSuccess success: @escaping (_ faceFilter: SCNNode) -> Void, onFailure failure: ((_ error: SvrfError) -> Void)? = nil) {
```
Comment (brentchow) [inline SvrfSDK/Source/SvrfSDK.swift]: Please update the README.md example

File: SvrfSDK/Source/SvrfError.swift (hunk 1 above)
Comment (AndreiEvstratenko) [reply SvrfSDK/Source/SvrfError.swift]: done

File: SvrfSDK/Source/SvrfSDK.swift [hunk 9]
```diff
@@ -35,36 +35,37 @@ public class SvrfSDK: NSObject {
...
      
      - parameters:
-       - onSuccess: Success closure.
-       - onFailure: Failure closure.
-       - error: Error message.
-    */
-    public static func authenticate(onSuccess success: @escaping () -> Void, onFailure failure: @escaping (_ error: SvrfError) -> Void) {
+     - onSuccess: Success closure.
+     - onFailure: Failure closure.
+     - error: Error message.
```
[the remark this answers is not in the document]
Comment (AndreiEvstratenko) [reply SvrfSDK/Source/SvrfSDK.swift]: done

File: SvrfSDK/Source/SvrfSDK.swift (hunk 3 above)
Comment (AndreiEvstratenko) [reply SvrfSDK/Source/SvrfSDK.swift]: done

File: SvrfSDK/Source/SvrfSDK.swift (hunk 2 above)
Comment (AndreiEvstratenko) [reply SvrfSDK/Source/SvrfSDK.swift]: done

File: SvrfSDK/Source/SvrfError.swift (hunk 4 above)
Comment (AndreiEvstratenko) [reply SvrfSDK/Source/SvrfError.swift]: done

File: SvrfSDK/Source/SvrfSDK.swift (hunk 8 above)
Comment (AndreiEvstratenko) [reply SvrfSDK/Source/SvrfSDK.swift]: done

File: SvrfSDK/Source/SvrfSDK.swift (hunk 7 above)
Comment (AndreiEvstratenko) [reply SvrfSDK/Source/SvrfSDK.swift]: I've updated examples and input parameters, did you mean that?

File: SvrfSDK/Source/SvrfSDK.swift (hunk 6 above)
Comment (AndreiEvstratenko) [reply SvrfSDK/Source/SvrfSDK.swift]: done

File: SvrfSDK/Source/SvrfSDK.swift (hunk 5 above)
Comment (AndreiEvstratenko) [reply SvrfSDK/Source/SvrfSDK.swift]: I've updated input parameters and examples, did you mean that?

File: SvrfSDK/Source/SvrfSDK.swift (hunk 7 above)
Comment (brentchow) [reply SvrfSDK/Source/SvrfSDK.swift]: yes. thank you

File: README.md [hunk 10]
```diff
@@ -223,10 +231,13 @@ Get a face filter *SCNNode* for *Media* with ID "547963".
 
 ```swift
 SvrfSDK.getMedia(id: "547963", onSuccess: { media in
-    let faceFilter = SvrfSDK.getFaceFilter(with: self.mtlDevice, media: media)
-    // Do what you want with the face filter
+    SvrfSDK.getFaceFilter(with: media, onSuccess: { faceFilter in
+        // Do what you want with the face filter
+    }, onFailure: { error in
```
Comment (ox) [inline README.md]: same question on failure closure as above ^

**Pull request merged**